### PR TITLE
nixos/networkd: add RoutingPolicyRules-related options

### PR DESCRIPTION
**Pull request by picnoir**
#### Motivation for this change
Mapping the systemd-networkd `RoutingPolicyRules` in the NixOS networkd module.

Check the `RoutingPolicyRule` section in the `systemd.network` for more informations.

###### Things done

- Added the relevant options in the module
- Modified the current networkd NixOS test to test all these options.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @flokli @fpletz @andir

### CODE REVIEW

File: nixos/modules/system/boot/networkd.nix [hunk 1]
```diff
@@ -535,6 +550,23 @@ let
     };
   };
 
+  routingPolicyRulesOptions = {
+    options = {
+      routingPolicyRuleConfig = mkOption {
+        default = { };
+        # TODO dd example
```
Comment (NinjaTrappeur) [inline nixos/modules/system/boot/networkd.nix]: Damn, forgot that. Fixing it ><

**Comment by NinjaTrappeur** [comment]
@GrahamcOfBorg test systemd-networkd

**Comment by NinjaTrappeur** [comment]
I'm puzzled by the ARM test error: https://logs.nix.ci/?key=nixos/nixpkgs.81405&attempt_id=adb7ba1f-8fb3-418c-af5a-68f0a2202fad

I do not have access to an interactive session on a aarch64 machine, it makes debugging this rather tricky :(

Looks like it's IPv6 related right?

File: nixos/modules/system/boot/networkd.nix [hunk 2]
```diff
@@ -535,6 +550,22 @@ let
     };
   };
 
+  routingPolicyRulesOptions = {
+    options = {
+      routingPolicyRuleConfig = mkOption {
+        default = { };
+        example = { routingPolicyRuleConfig = { Table = 10; IncomingInterface = "eth1"; } ;};
```
Comment (mweinelt) [inline nixos/modules/system/boot/networkd.nix]: Maybe add a dualstacked example by setting `Family=both`, because this rule would otherwise be IPv4-only.

**Comment by NinjaTrappeur** [comment]
@GrahamcOfBorg test systemd-networkd

**Comment by NinjaTrappeur** [comment]
Aha! Found the root cause of the ARM failure: we're missing the `CONFIG_IP_MULTIPLE_TABLES` kernel config flag.

**Comment by NinjaTrappeur** [comment]
@GrahamcOfBorg test systemd-networkd

**Comment by NinjaTrappeur** [comment]
The aarch test fails because of #61602. I think it's safe to ignore it for now.

**Comment by flokli** [comment]
Rebased on top of the landed aarch64 fix in master.

@GrahamcOfBorg test systemd-networkd

**Review by flokli — APPROVED** [review]

File: nixos/modules/system/boot/networkd.nix [hunk 3]
```diff
@@ -235,6 +235,21 @@ let
...
   ];
 
+  checkRoutingPolicyRule = checkUnitConfig "RoutingPolicyRule" [
+    (assertOnlyFields [
+      "TypeOfService" "From" "To" "FirewallMark" "Table" "Priority"
+      "IncomingInterface" "OutgoingInterface" "SourcePort" "DestinationPort"
+      "IPProtocol" "InvertRule" "Family"
+    ])
+    (assertRange "TypeOfService" 0 255)
+    (assertRange "FirewallMark" 1 4294967295)
```
Comment (Ma27) [inline nixos/modules/system/boot/networkd.nix]: Hmm doesn't this break eval with Nix <2.2 on 32bit machines? See also #45722, the issue has been fixed in Nix in https://github.com/NixOS/nix/pull/2378, but AFAICS this is only part of Nix 2.2 (and according to `lib/minver.nix` we support nixpkgs evaluation with at least Nix 2.0).

**Review by Ma27 — COMMENTED** [review]
Apart from that one comment, LGTM :+1: - thanks!

File: nixos/modules/system/boot/networkd.nix (hunk 3 above)
Comment (NinjaTrappeur) [reply nixos/modules/system/boot/networkd.nix]: Hah! Did not know about that.

Removing the check then :(

**Comment by NinjaTrappeur** [comment]
Updated. I also removed some other 64bit integer occurrences in the process.

Naive question: Nix 2.2 was release more than a year ago. Any reason to keep supporting Nix < 2.2?

@GrahamcOfBorg test systemd-networkd

**Comment by Ma27** [comment]
> Naive question: Nix 2.2 was released more than a year ago. Any reason to keep supporting Nix < 2.2?

Not sure, what do others think of it?

**Review by Ma27 — APPROVED** [review]

**Comment by andir** [comment]
On 03:22 04.03.20, Maximilian Bosch wrote:
> > Naive question: Nix 2.2 was released more than a year ago. Any reason to keep supporting Nix < 2.2?
> 
> Not sure, what do others think of it?

IIRC we had that topic on IRC a few times. The conclusion was something
like we should be ~2y backwards compat. I am not sure how much of a
policy that is and who said it…

**Pull request merged**